File: native/engine/android/app/src/com/brix/jsb/JsbCall.java
/*
 * @Date: 2024-02-26 10:53:25
 * @LastEditors: zhaozhenguo zhaozhenguo@tuyoogame.com
 * @LastEditTime: 2024-02-27 14:00:39
 */
package com.brix.jsb;

import android.util.Log;

import com.cocos.lib.CocosHelper;
import com.cocos.lib.CocosJavascriptJavaBridge;
import com.cocos.lib.JsbBridge;

import org.json.JSONException;
import org.json.JSONObject;


public class JsbCall {

    static final private String TAG = "[JsbCall]";

    public static void exec(String service, String action, String params, String callbackId) {
        Log.d(TAG, "exec: " + service + " " + action + " " + params + " " + callbackId);
        final CallbackContext callbackContext = new CallbackContext(callbackId);
        try {
            JsbCall.exec(service, action, new JSONObject(params), callbackContext);
        } catch (JSONException e) {
            Log.e(TAG, "exec: e: ", e);
            callbackContext.failure(String.format("params to json or params error, params: %s", params));
        }
    }


    private static boolean exec(String service, String action, JSONObject args, final CallbackContext callbackContext) throws JSONException {
        return PluginManager.getInstance().exec(service, action, args, callbackContext);
    }

    public static void emitWindowEvent(final String event, final JSONObject data) {
        // Log.d(TAG, "emitWindowEvent: " + event + " data: " + data.toString());
        CocosHelper.runOnGameThread(new Runnable() {
            @Override
            public void run() {
//                String dataStr = "";
//                try{
//                   dataStr = URLEncoder.encode(data.toString(), "UTF-8");
//                }catch (UnsupportedEncodingException e){
//                    dataStr = "";
//                }
                String dataStr = "";

                String jsStr = String.format("window.JsbNativeCall.emitWindowEvent('%s', '%s')", event, dataStr);
                // Log.d(TAG, "emitWindowEvent: jsStr:" + jsStr);
                CocosJavascriptJavaBridge.evalString(jsStr);
            }
        });
    }
    public static void emitIapEvent(String eventName, JSONObject eventData) {
        Log.d(TAG, "emitWindowEvent[iap]: " + eventName + " data: " + eventData.toString());
        CocosHelper.runOnGameThread(new Runnable() {
            @Override
            public void run() {
                String dataStr =eventData==null?"{}": eventData.toString();
                String jsStr = String.format("window.JsbNativeCall.emitWindowEvent('%s', '%s')", eventName, dataStr);
//                Log.d(TAG, "emitWindowEvent: jsStr:" + jsStr);
                CocosJavascriptJavaBridge.evalString(jsStr);
            }
        });
    }


    public static void successCall(final String callbackId, final String params) {
        Log.d(TAG, "successCall: " + callbackId + " params: " + params);
        CocosHelper.runOnGameThread(new Runnable() {
            @Override
            public void run() {
                // Log.d(TAG, "successCall: log....");
                String jsStr = String.format("window.JsbNativeCall.callBackCallSuccess('%s', '%s')", callbackId, params);
                // Log.d(TAG, "successCall: evalString: " + jsStr);
                CocosJavascriptJavaBridge.evalString(jsStr);
            }
        });
    }

    public static void failureCall(final String callbackId, final String params) {
        Log.d(TAG, "failureCall: " + callbackId + " params: " + params);

        CocosHelper.runOnGameThread(new Runnable() {
            @Override
            public void run() {
                String jsStr = String.format("window.JsbNativeCall.callBackCallFailure('%s', '%s')", callbackId, params);
                Log.d(TAG, "failureCall: evalString: " + jsStr);
                CocosJavascriptJavaBridge.evalString(jsStr);
            }
        });
    }
}

File: native/engine/ios/Plugins/JsbCall.m
// JsbCall.m

#import "JsbCall.h"
#include "CallbackContext.h"
#include "PluginManager.h"
#include "platform/apple/JsbBridgeWrapper.h"
//#include "application/ApplicationManager.h"
//#include "cocos/bindings/jswrapper/SeApi.h"
@implementation JsbCall

+ (void)init{
    JsbBridgeWrapper* m = [JsbBridgeWrapper sharedInstance];
      OnScriptEventListener jsbCallListener = ^void(NSString* args){
          NSData *data = [args dataUsingEncoding:NSUTF8StringEncoding];
          NSError *error = nil;
          NSDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];

          if (jsonObject) {
              // 从 JSON 对象中获取参数值
              NSString *service = jsonObject[@"service"];
              NSString *action = jsonObject[@"action"];
              NSString *params = jsonObject[@"params"];
              NSString *callbackId = jsonObject[@"callbackId"];
              [JsbCall execWithService:service action:action params:params callbackId:callbackId];
          } else {
              NSLog(@"Error parsing JSON: %@", error.localizedDescription);
          }
      };
    [m addScriptEventListener:@"JsbCall" listener:jsbCallListener];
}
+ (void)execWithService:(NSString *)service action:(NSString *)action params:(NSString *)params callbackId:(NSString *)callbackId {
    NSLog(@"exec: %@ %@ %@ %@", service, action, params, callbackId);
    // 在这里调用相应的方法，并根据结果调用成功或失败的回调
    CallbackContext *callbackContext = [[CallbackContext alloc] initWithCallbackId:callbackId];
    NSData *jsonData = [params dataUsingEncoding:NSUTF8StringEncoding];
    NSError *error;
    id jsonObject = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:&error];
    NSDictionary *resultDictionary = (NSDictionary *)jsonObject;
    [[PluginManager sharedInstance] execWithService:service action:action args:resultDictionary callback:callbackContext];
}

+ (void)emitWindowEvent:(NSString *)event data:(NSDictionary *)data {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSDictionary *jsonObject = @{@"event": event, @"data": data};
        JsbBridgeWrapper* m = [JsbBridgeWrapper sharedInstance];
              [m dispatchEventToScript:@"emitWindowEvent" arg:[self jsonStringFromDictionary:jsonObject]];
    });
}

+ (void)emitIapEvent:(NSString *)eventName eventData:(NSDictionary *)eventData {
    NSLog(@"emitWindowEvent[iap]: %@ data: %@", eventName, eventData);
    // 在这里调用 emitWindowEvent 方法
    [self emitWindowEvent:eventName data:eventData];
}

+ (void)successCall:(NSString *)callbackId params:(NSString *)params {
    NSLog(@"successCall: %@ params: %@", callbackId, params);
    // 在这里调用相应的方法，并根据结果调用成功或失败的回调
    NSDictionary *jsonObject = @{@"event": callbackId, @"data": params};
    JsbBridgeWrapper* m = [JsbBridgeWrapper sharedInstance];
          [m dispatchEventToScript:@"callBackCallSuccess" arg:[self jsonStringFromDictionary:jsonObject]];
}

+ (void)failureCall:(NSString *)callbackId params:(NSString *)params {
    NSLog(@"failureCall: %@ params: %@", callbackId, params);
    // 在这里调用相应的方法，并根据结果调用成功或失败的回调
    NSDictionary *jsonObject = @{@"event": callbackId, @"data": params};
    JsbBridgeWrapper* m = [JsbBridgeWrapper sharedInstance];
          [m dispatchEventToScript:@"callBackCallFailure" arg:[self jsonStringFromDictionary:jsonObject]];
}

+ (NSString *)jsonStringFromDictionary:(NSDictionary *)dictionary {
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dictionary options:0 error:&error];
    if (!jsonData) {
        NSLog(@"jsonStringFromDictionary error: %@", error.localizedDescription);
        return @"{}";
    } else {
        return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    }
}

@end
